src/core/replication_fns.rs

```diff
@@ -16,160 +20,329 @@ pub struct ReplicationFns {
+    ///
+    /// Should be used after calling
+    /// [`CommandMarkers::insert`](super::command_markers::CommandMarkers::insert)
+    pub(super) fn register_marker(&mut self, marker_id: CommandMarkerId) {
+        self.marker_slots += 1;
+        for (command_fns, _) in &mut self.commands {
+            command_fns.add_marker_slot(marker_id);
+        }
+    }
+
+    /// Associates command functions with a marker.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that passed `write` can be safely called with a
+    /// [`SerdeFns`] created for `C`.
     ///
-    /// Returned ID can be assigned to a component inside
-    /// [`ReplicationRule`](super::replication_rules::ReplicationRule).
+    /// # Panics
     ///
-    /// Could be called multiple times for the same component with different functions.
-    pub fn register_component_fns(&mut self, fns: ComponentFns) -> ComponentFnsId {
-        self.components.push(fns);
+    /// Panics if marker wasn't registered. Use [`Self::register_marker`] first.
+    pub(super) unsafe fn register_marker_fns<C: Component>(
+        &mut self,
+        world: &mut World,
+        marker_id: CommandMarkerId,
+        write: WriteFn,
+        remove: RemoveFn,
+    ) {
+        let (index, _) = self.init_command_fns::<C>(world);
+
+        // SAFETY: index obtained from `Self::init_command_fns` is always valid.
+        let (command_fns, _) = self.commands.get_unchecked_mut(index);
 
-        ComponentFnsId(self.components.len() - 1)
+        // SAFETY: `command_fns` was created for `C` and the caller ensured
+        // that `write` can be safely called with a `SerdeFns` created for `C`.
+        command_fns.set_marker_fns(marker_id, write, remove);
     }
 
-    /// Returns a reference to registered component functions.
+    /// Same as [`Self::register_serde_fns`], but uses default functions for a component.
     ///
-    /// # Panics
+    /// If your component contains any [`Entity`] inside, use [`Self::register_default_serde_fns`].
+    pub fn register_default_serde_fns<C>(&mut self, world: &mut World) -> FnsInfo
+    where
+        C: Component + Serialize + DeserializeOwned,
+    {
+        self.register_serde_fns(
+            world,
+            serde_fns::serialize::<C>,
+            serde_fns::deserialize::<C>,
+            serde_fns::deserialize_in_place::<C>,
+        )
+    }
+
+    /// Same as [`Self::register_serde_fns`], but uses default functions for a mapped component.
+    ///
+    /// Always use it for components that contain entities.
+    ///
+    /// See also [`Self::register_default_serde_fns`].
+    pub fn register_mapped_serde_fns<C>(&mut self, world: &mut World) -> FnsInfo
+    where
+        C: Component + Serialize + DeserializeOwned + MapEntities,
+    {
+        self.register_serde_fns(
+            world,
+            serde_fns::serialize::<C>,
+            serde_fns::deserialize_mapped::<C>,
+            serde_fns::deserialize_in_place::<C>,
```

Comment:
I decided to go with `in_place_as_deserialize` since we are not assigning actually.
But let me know if you have other suggestions.

I also pushed a small rename of other default functions. It should be more clear now.